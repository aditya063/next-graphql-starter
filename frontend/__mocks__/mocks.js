export const user = {
  id: '1',
  email: 'a@a.com'
};

export const idea = {
  id: '1',
  content: 'mock content'
};

export const newIdea = {
  id: '1',
  content: 'new mock content'
};

export const users = [user];
export const ideas = [user];
