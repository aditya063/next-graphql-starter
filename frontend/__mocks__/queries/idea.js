import * as query from '../../graphql/queries/idea';

export const MOCK_ME_IDEAS_QUERY = {
  request: { query: query.ME_IDEAS_QUERY },
  result: {
    data: {
      getUserIdeas: [
        { id: '1', content: 'mock idea 1 content' },
        { id: '2', content: 'mock idea 2 content' }
      ]
    }
  }
};

export const MOCK_ERROR_ME_IDEAS_QUERY = {
  request: { query: query.ME_IDEAS_QUERY },
  error: new Error('mock error')
};

export const MOCK_CREATE_IDEA_MUTATION = {
  request: {
    query: CREATE_IDEA_MUTATION,
    variables: { content: 'mock created idea' }
  },
  result: { data: { createIdea: [{ id: '1' }] } }
};

export const MOCK_UPDATE_IDEA_MUTATION = {
  request: { query: UPDATE_IDEA_MUTATION },
  result: {
    data: { updateIdea: { id: '2', content: 'updated mock idea 2 content' } }
  }
};

export const MOCK_DELETE_IDEA_MUTATION = {
  request: { query: DELETE_IDEA_MUTATION, variables: { id: '1' } },
  result: { data: { deleteIdea: { id: '1' } } }
};
