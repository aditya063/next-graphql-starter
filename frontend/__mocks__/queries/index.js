export {
  MOCK_IS_LOGGED_IN,
  MOCK_CURRENT_USER,
  MOCK_ERROR_CURRENT_USER,
  MOCK_SIGN_UP,
  MOCK_SIGN_IN,
  MOCK_SIGN_OUT,
  MOCK_REQUEST_RESET,
  MOCK_RESET_PASSWORD
} from './user';

export {
  MOCK_CREATE_IDEA,
  MOCK_UPDATE_IDEA,
  MOCK_DELETE_IDEA
} from './idea';
