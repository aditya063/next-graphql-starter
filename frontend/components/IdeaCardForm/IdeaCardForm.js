import { Mutation } from 'react-apollo';

import {
  CURRENT_USER_PAGINATED_IDEAS,
  CREATE_IDEA_MUTATION
} from '../../graphql/queries';
import { pageSize } from '../../config';
import * as sc from './IdeaCardForm.style';

class IdeaCardForm extends React.PureComponent {
  state = { idea: '', isSubmitDisabled: true };

  canSubmit = () => {
    if (this.state.idea === '') {
      this.setState({ isSubmitDisabled: true });
    } else {
      this.setState({ isSubmitDisabled: false });
    }
  };

  handleChangeIdeaInput = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value }, this.canSubmit);
  };

  handleSubmitIdeaForm = async (event, createIdea) => {
    event.preventDefault();
    this.setState({ isSubmitDisabled: true });
    createIdea();
    this.setState({ idea: '' }, this.canSubmit);
  };

  handleError = error => error;

  handleUpdate = (cache, data) => {
    // Read the data from cache for this query
    const ideasData = cache.readQuery({
      query: CURRENT_USER_PAGINATED_IDEAS,
      variables: { orderBy: 'createdAt_DESC', first: pageSize }
    });

    const newIdeas = ideasData.currentUserPaginatedIdeas.edges;

    // Add idea from the mutation to the beginning
    newIdeas.unshift({
      node: { ...data.data.createIdea },
      __typename: 'IdeaEdge'
    });

    // Write data back to the cache
    cache.writeQuery({
      query: CURRENT_USER_PAGINATED_IDEAS,
      variables: { orderBy: 'createdAt_DESC', first: pageSize },
      data: {
        ...ideasData,
        currentUserPaginatedIdeas: {
          ...ideasData.currentUserPaginatedIdeas,
          edges: newIdeas
        }
      }
    });
  };

  render() {
    return (
      <Mutation
        mutation={CREATE_IDEA_MUTATION}
        variables={{ content: this.state.idea }}
        onError={this.handleError}
        update={this.handleUpdate}
      >
        {createIdea => (
          <sc.form
            onSubmit={event => this.handleSubmitIdeaForm(event, createIdea)}
          >
            <img src="static/ideabox.png" alt="ideabox" />

            <textarea
              name="idea"
              type="text"
              placeholder="What's on your mind?"
              value={this.state.idea}
              onChange={event => this.handleChangeIdeaInput(event)}
            />

            <button type="submit" disabled={this.state.isSubmitDisabled}>
              Add Idea
            </button>
          </sc.form>
        )}
      </Mutation>
    );
  }
}

export default IdeaCardForm;
