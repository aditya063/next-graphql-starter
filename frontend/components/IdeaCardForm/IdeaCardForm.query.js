import gql from 'graphql-tag';
import { ME_IDEAS_QUERY } from '../IdeaCardContainer/IdeaCardContainer.query';

export { ME_IDEAS_QUERY };

export const CREATE_IDEA_MUTATION = gql`
  mutation CREATE_IDEA_MUTATION($content: String!) {
    createIdea(content: $content) {
      id
    }
  }
`;

export const MOCK_CREATE_IDEA_MUTATION = {
  request: {
    query: CREATE_IDEA_MUTATION
  },
  result: {
    data: {
      createIdea: [
        {
          id: '1'
        }
      ]
    }
  }
};
