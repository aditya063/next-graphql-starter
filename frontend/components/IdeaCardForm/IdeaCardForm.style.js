import styled from 'styled-components';
import { Form, Input, Button } from 'antd';

export const IdeaCardForm = styled(Form)`
  display: flex;
  position: relative;
`;

export const BoxImg = styled.img`
  position: absolute;
  left: 10px;
  top: -30px;
  width: 50px;
`;

export const InputTextArea = styled(Input.TextArea)`
  margin-bottom: 0 !important;
  border: none;
  border-right: 1px solid ${props => props.theme.color.black};

  &:focus {
    box-shadow: none;
  }
`;

export const SubmitBtn = styled(Button)`
  height: auto;
  border: none;
`;
