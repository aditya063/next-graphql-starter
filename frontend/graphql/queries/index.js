export {
  IS_LOGGED_IN,
  CURRENT_USER_QUERY,
  SIGN_UP_MUTATION,
  SIGN_IN_MUTATION,
  SIGN_OUT_MUTATION,
  REQUEST_RESET_MUTATION,
  RESET_PASSWORD_MUTATION
} from './user';

export {
  CREATE_IDEA_MUTATION,
  UPDATE_IDEA_MUTATION,
  DELETE_IDEA_MUTATION
} from './idea';
