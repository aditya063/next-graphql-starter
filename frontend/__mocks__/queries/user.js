import * as query from '../../graphql/queries/user';

export const MOCK_IS_LOGGED_IN = {
  request: { query: query.IS_LOGGED_IN },
  result: { data: { isLoggedIn: true } }
};

export const MOCK_CURRENT_USER = {
  request: { query: query.CURRENT_USER },
  result: {
    data: {
      currentUser: { id: '1', email: 'mock@email.com', color: 'mock color' }
    }
  }
};

export const MOCK_ERROR_CURRENT_USER = {
  request: { query: query.CURRENT_USER },
  error: new Error('mock error')
};

export const MOCK_SIGN_UP = {
  request: {
    query: query.SIGN_UP,
    variables: {
      email: 'mock@email.com',
      password: 'mockpass',
      confirmPassword: 'confirmed mockpass'
    }
  },
  result: { data: { signUp: [{ id: '1' }] } }
};

export const MOCK_SIGN_IN = {
  request: {
    query: query.SIGN_IN,
    variables: { email: 'mock@email.com', password: 'mockpass' }
  },
  result: { data: { signIn: [{ id: '1' }] } }
};

export const MOCK_SIGN_OUT = {
  request: { query: query.SIGN_OUT },
  result: { data: { signOut: true } }
};

export const MOCK_REQUEST_RESET = {
  request: {
    query: query.REQUEST_RESET,
    variables: { email: 'mock@email.com' }
  },
  result: { data: { requestReset: true } }
};

export const MOCK_RESET_PASSWORD = {
  request: {
    query: query.RESET_PASSWORD,
    variables: {
      resetToken: '4e4a8fb6e44ec32642cfa410243652f85885bc72',
      password: 'mockpass',
      confirmPassword: 'mockpass'
    }
  },
  result: { data: { resetPassword: [{ id: '1', email: 'mock@email.com' }] } }
};
