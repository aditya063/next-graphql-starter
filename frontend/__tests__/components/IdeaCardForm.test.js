import { mount } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';
import { ThemeProvider } from 'styled-components';

import { load } from '../../utils/testing';
import IdeaCardForm from '../../components/IdeaCardForm/IdeaCardForm';
import { MOCK_CREATE_IDEA } from '../../__mocks__/queries';
import theme from '../../styles/theme.style';

describe('IdeaCardForm', () => {
  let mockProps;
  let mockQueries;
  let wrapper;

  beforeEach(() => {
    jest.resetAllMocks();
    mockProps = {};
    mockQueries = [MOCK_CREATE_IDEA];
    wrapper = mount(
      <MockedProvider mocks={mockQueries} addTypename={false}>
        <ThemeProvider theme={theme}>
          <IdeaCardForm {...mockProps} />
        </ThemeProvider>
      </MockedProvider>,
      {
        disableLifecycleMethods: true
      }
    );
  });

  it('matches snapshot', () => {
    const wrapSnap = wrapper.find('IdeaCardFormstyle__form');

    // console.log(wrapper.debug());

    expect(wrapSnap).toMatchSnapshot();
  });

  it('matches snapshot - loaded', async () => {
    const wrapSnap = wrapper.find('IdeaCardFormstyle__form');

    await load(wrapper);

    // console.log(wrapper.debug());

    expect(wrapSnap).toMatchSnapshot();
  });
});
