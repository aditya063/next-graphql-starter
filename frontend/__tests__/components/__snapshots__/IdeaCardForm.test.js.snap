// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdeaCardForm matches snapshot - loaded 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  grid-column: span 2;
  position: relative;
}

.c0 > img {
  width: 50px;
  position: absolute;
  top: -30px;
  left: 10px;
}

.c0 > textarea {
  padding: 10px;
  outline: none;
  min-height: 56px;
  width: 100%;
  resize: vertical;
  border: none;
  margin-right: 1px;
}

.c0 > button {
  border: none;
  outline: none;
  cursor: pointer;
  background-color: #fffacd;
  padding: 0;
}

.c0 > button:active {
  background-color: #FFFF99;
}

@media screen and (min-width:600px) {
  .c0 > img {
    left: 0;
  }
}

<IdeaCardFormstyle__form
  onSubmit={[Function]}
>
  <StyledComponent
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "IdeaCardFormstyle__form-sc-4er11-0",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "display:flex;width:100%;grid-column:span 2;position:relative;> img{width:50px;position:absolute;top:-30px;left:10px;}> textarea{padding:10px;outline:none;min-height:56px;width:100%;resize:vertical;border:none;margin-right:1px;}> button{border:none;outline:none;cursor:pointer;background-color:",
            [Function],
            ";padding:0;&:active{background-color:",
            [Function],
            ";}}@media screen and (min-width:600px){> img{left:0;}}",
          ],
        },
        "displayName": "IdeaCardFormstyle__form",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "IdeaCardFormstyle__form-sc-4er11-0",
        "target": "form",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    onSubmit={[Function]}
  >
    <form
      className="c0"
      onSubmit={[Function]}
    >
      <img
        alt="ideabox"
        src="static/ideabox.png"
      />
      <textarea
        name="idea"
        onChange={[Function]}
        placeholder="What's on your mind?"
        type="text"
        value=""
      />
      <button
        disabled={true}
        type="submit"
      >
        Add Idea
      </button>
    </form>
  </StyledComponent>
</IdeaCardFormstyle__form>
`;

exports[`IdeaCardForm matches snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  grid-column: span 2;
  position: relative;
}

.c0 > img {
  width: 50px;
  position: absolute;
  top: -30px;
  left: 10px;
}

.c0 > textarea {
  padding: 10px;
  outline: none;
  min-height: 56px;
  width: 100%;
  resize: vertical;
  border: none;
  margin-right: 1px;
}

.c0 > button {
  border: none;
  outline: none;
  cursor: pointer;
  background-color: #fffacd;
  padding: 0;
}

.c0 > button:active {
  background-color: #FFFF99;
}

@media screen and (min-width:600px) {
  .c0 > img {
    left: 0;
  }
}

<IdeaCardFormstyle__form
  onSubmit={[Function]}
>
  <StyledComponent
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "IdeaCardFormstyle__form-sc-4er11-0",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "display:flex;width:100%;grid-column:span 2;position:relative;> img{width:50px;position:absolute;top:-30px;left:10px;}> textarea{padding:10px;outline:none;min-height:56px;width:100%;resize:vertical;border:none;margin-right:1px;}> button{border:none;outline:none;cursor:pointer;background-color:",
            [Function],
            ";padding:0;&:active{background-color:",
            [Function],
            ";}}@media screen and (min-width:600px){> img{left:0;}}",
          ],
        },
        "displayName": "IdeaCardFormstyle__form",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "IdeaCardFormstyle__form-sc-4er11-0",
        "target": "form",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    onSubmit={[Function]}
  >
    <form
      className="c0"
      onSubmit={[Function]}
    >
      <img
        alt="ideabox"
        src="static/ideabox.png"
      />
      <textarea
        name="idea"
        onChange={[Function]}
        placeholder="What's on your mind?"
        type="text"
        value=""
      />
      <button
        disabled={true}
        type="submit"
      >
        Add Idea
      </button>
    </form>
  </StyledComponent>
</IdeaCardFormstyle__form>
`;
